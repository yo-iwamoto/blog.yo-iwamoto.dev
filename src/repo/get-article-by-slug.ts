import { getServerSideEnv } from "@/config/env";
import { transformArticleHtml } from "@/lib/transform-article-html";
import { Article } from "@/model/article";
import { cache } from "react";
import { cdnClient } from "./client";
import type { Content } from "./client";

const articleModelUid = "article";

async function getArticleBySlugFn(slug: string): Promise<Article | null> {
  const res = await cdnClient.getFirstContent<
    {
      title: string;
      slug: string;
      tags: { name: string; slug: string }[];
      body: string;
    } & Content
  >({
    appUid: getServerSideEnv().newtAppUid,
    modelUid: articleModelUid,
    query: {
      slug,
    },
  });
  if (res === null) return null;

  const article = new Article(
    res.title,
    res.slug,
    res.tags,
    new Date(res._sys.raw.firstPublishedAt),
    transformArticleHtml(res.body),
  );

  return article;
}

export const getArticleBySlug = cache(getArticleBySlugFn);
