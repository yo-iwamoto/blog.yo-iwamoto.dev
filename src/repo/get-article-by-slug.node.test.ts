import { getArticleBySlug } from "./get-article-by-slug";

const { getFirstContent } = vi.hoisted(() => ({
  getFirstContent: vi.fn(async () => ({
    title: "Article 1",
    slug: "article-1",
    tags: [],
    body: "Hello World",
    _sys: {
      raw: {
        firstPublishedAt: "2021-01-01T00:00:00.000Z",
      },
    },
  })),
}));

vi.mock("./client", () => ({ cdnClient: { getFirstContent } }));

vi.mock("@/lib/transfrom-article-html", () => ({
  transformArticleHtml: (body: string) => body,
}));

describe("getArticleBySlug", () => {
  it("取得した記事が整形されて返却されること", async () => {
    const res = await getArticleBySlug("article-1");

    expect(res?.title).toBe("Article 1");
    expect(res?.slug).toBe("article-1");
    expect(res?.tags).toHaveLength(0);
    expect(res?.publishedAt).toStrictEqual(
      new Date("2021-01-01T00:00:00.000Z"),
    );
    expect(res?.body).toBe("Hello World");
  });

  it("クエリが変わらないこと", async () => {
    await getArticleBySlug("article-1");

    expect(getFirstContent.mock.lastCall).toMatchSnapshot();
  });
});
