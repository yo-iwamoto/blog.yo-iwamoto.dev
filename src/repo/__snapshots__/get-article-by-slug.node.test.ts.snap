// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getArticleBySlug > クエリが変わらないこと 1`] = `
[
  {
    "appUid": "xxx",
    "modelUid": "article",
    "query": {
      "slug": "article-1",
    },
  },
]
`;
